core/src/main/scala/org/apache/spark/h2o/backends/internal/H2ORpcEndpoint.scala

```diff
@@ -48,17 +49,34 @@ class H2ORpcEndpoint(override val rpcEnv: RpcEnv) extends ThreadSafeRpcEndpoint
 
     case StopEndpointMsg =>
       this.stop()
+    case LockClusterMsg =>
+      Paxos.lockCloud("Locking the cloud from Sparkling Water as we have reached the expected cluster size.")
   }
 
   override def receiveAndReply(context: RpcCallContext): PartialFunction[Any, Unit] = {
     case StartH2OWorkersMsg(conf) =>
       val nodeDesc = InternalH2OBackend.startH2OWorker(conf)
       context.reply(nodeDesc)
+    case CheckClusterSizeMsg =>
+      context.reply(H2O.CLOUD.size())
+    case IsLeaderNodeMsg =>
+      val reply = if (H2O.SELF.isLeaderNode) {
+        Some(NodeDesc(SparkEnv.get.executorId, H2O.SELF_ADDRESS.getHostAddress, H2O.API_PORT))
+      } else {
+        None
+      }
+      context.reply(reply)
   }
 }
 
-case class StopEndpointMsg()
+case object StopEndpointMsg
 
 case class FlatFileMsg(nodes: Array[NodeDesc], portOffset: Int)
 
 case class StartH2OWorkersMsg(conf: H2OConf)
+
+case object CheckClusterSizeMsg
+
+case object LockClusterMsg
+
+case object IsLeaderNodeMsg
```

Comment:
nit: if `Option[NodeDesc]` is returned shouldn't it be renamed to `getLeaderNodeMsg` or similar?

Reply:
Good point, agree